node/actors/bft/src/testonly/ut_harness.rs

```diff
@@ -0,0 +1,353 @@
+    }
+
+    pub(crate) async fn new_procedural_leader_commit(&mut self) -> Signed<ConsensusMsg> {
+        let replica_prepare = self.new_current_replica_prepare(|_| {});
+        self.dispatch_replica_prepare(replica_prepare.clone())
+            .unwrap();
+        let leader_prepare = self.recv_signed().await.unwrap();
+        self.dispatch_leader_prepare(leader_prepare).await.unwrap();
+        let replica_commit = self.recv_signed().await.unwrap();
+        self.dispatch_replica_commit(replica_commit).unwrap();
+        self.recv_signed().await.unwrap()
+    }
+
+    #[allow(clippy::result_large_err)]
+    pub(crate) fn dispatch_replica_prepare_many(
+        &mut self,
+        messages: Vec<ReplicaPrepare>,
+        keys: Vec<SecretKey>,
+    ) -> Result<(), ReplicaPrepareError> {
+        messages
+            .iter()
+            .enumerate()
+            .map(|(i, msg)| {
+                let signed = keys[i].sign_msg(ConsensusMsg::ReplicaPrepare(msg.clone()));
+                self.dispatch_replica_prepare(signed)
+            })
+            .last()
+            .unwrap()
+    }
+
+    #[allow(clippy::result_large_err)]
+    pub(crate) fn dispatch_replica_prepare(
+        &mut self,
+        msg: Signed<ConsensusMsg>,
+    ) -> Result<(), ReplicaPrepareError> {
+        self.consensus.leader.process_replica_prepare(
+            &self.ctx,
+            &self.consensus.inner,
+            msg.cast().unwrap(),
+        )
+    }
+
+    pub(crate) fn dispatch_replica_commit(
+        &mut self,
+        msg: Signed<ConsensusMsg>,
+    ) -> Result<(), ReplicaCommitError> {
+        self.consensus.leader.process_replica_commit(
+            &self.ctx,
+            &self.consensus.inner,
+            msg.cast().unwrap(),
+        )
+    }
+
+    pub(crate) async fn dispatch_leader_prepare(
+        &mut self,
+        msg: Signed<ConsensusMsg>,
+    ) -> Result<(), LeaderPrepareError> {
+        scope::run!(&self.ctx, |ctx, s| {
+            s.spawn(async {
+                let res = self
+                    .consensus
+                    .replica
+                    .process_leader_prepare(ctx, &self.consensus.inner, msg.cast().unwrap())
+                    .await;
+                Ok(res)
+            })
+            .join(ctx)
+        })
+        .await
+        .unwrap()
+    }
+
+    pub(crate) async fn dispatch_leader_commit(
+        &mut self,
+        msg: Signed<ConsensusMsg>,
+    ) -> Result<(), LeaderCommitError> {
+        scope::run!(&self.ctx, |ctx, s| {
+            s.spawn(async {
+                let res = self
+                    .consensus
+                    .replica
+                    .process_leader_commit(ctx, &self.consensus.inner, msg.cast().unwrap())
+                    .await;
+                Ok(res)
+            })
+            .join(ctx)
+        })
+        .await
+        .unwrap()
+    }
+
+    pub(crate) async fn recv_signed(&mut self) -> Option<Signed<ConsensusMsg>> {
```

Comment:
This method seems to always return `Some`; could it be simplified to return `Signed<_>` instead of an `Option`?

Reply:
Yes, originally it had `None` which turned out to be unnecessary, but the method signature remained. 

I changed it now to return `Result` which will propagate the error type from `self.pipe.recv` instead of just unwrapping it: https://github.com/matter-labs/era-consensus/pull/39/commits/aa85e36a11d004eae9ceea493a43bd1861574ae8